File: DS-and-Algorithms/Coderust/Arrays/1-Binary-Search-Sorted-Array/binary_search.go
package main
import (
	"fmt"
	"strconv"
)

func BinarySearchRecursive(arr []int, l int, r int, x int) int {
	if r >= l {
		mid := l + (r-l)/2
		if arr[mid] == x {
			return mid
		}
		if arr[mid] > x {
			return BinarySearchRecursive(arr, l, mid-1, x)
		}
		return BinarySearchRecursive(arr, mid+1, r, x)
	}
	return -1
}

func BinarySearch(nums []int, target int) int {
	return BinarySearchRec(nums, target, 0, len(nums)-1)
}

/* ArrayToString is used to convert a string array to string.
ArrayToString is used as an helper function in a main function for
printing purposes. */
func ArrayToString(nums []int) string{
	if len(nums) == 0 {
		return "[]"
	}
	res := "["
	for _, num := range nums {
		res += strconv.Itoa(num) + ", "
	}
	res += "]"
	return res[:len(res) - 3] + "]"
}

func main() {
	numsLists := [][]int{
		{},
		{0, 1},
		{1, 2, 3},
		{-1, 0, 3, 5, 9, 12},
		{-1, 0, 3, 5, 9, 12},
	}

	targetList := []int{12, 1, 3, 9, 2}
	for i, nums := range numsLists {
		target := targetList[i]
		index := BinarySearch(nums, target)
		fmt.Printf("%d. Array to search: %s\n", i+1, ArrayToString(nums))
		fmt.Printf("   Target: %d\n", target)
		if index != -1 {
			fmt.Printf("   %d exists in the array at index %d\n", target, index)
		} else {
			fmt.Printf("   %d does not exist in the array so the return value is %d\n", target, index)
		}
		fmt.Printf(
			"----------------------------------------------------------------------------------------------------\n\n",
		)
	}
}

File: DS-and-Algorithms/Coderust/Arrays/1-Binary-Search-Sorted-Array/binary_search.rb
# Time complexity is O(log(n))
# Space complexity is O(1)

def binary_search(numbers, target)
  low = 0
  high = numbers.length - 1
  while low <= high
    mid = (low + high) / 2
    if numbers[mid] == target
      return mid
    elsif numbers[mid] > target
      high = mid - 1
    else
      low = mid + 1
    end
  end
  return -1
end

nums_lists = [[], [0,1], [1,2,3], [-1,0,3,5,9,12], [-1,0,3,5,9,12]]
target_list = [12, 1, 3, 9, 2]
for i in 0..nums_lists.length-1
	nums = nums_lists[i]
	target = target_list[i]
	index = binary_search(nums, target)
	if index != -1
		print(target.to_s + " exists in the array and its index is ", index.to_s)
	else
		print(target.to_s, " does not exist in the array so the return value is ", index.to_s)
	end
	print("\n----------------------------------------------------------------------------------------------------\n\n")
end

# Recursive
# Time complexity is O(log(n))
# Space complexity is O(log(n))

def binary_search(numbers, target)
  return -1 if numbers.length == 0
  mid = numbers.length / 2
  if numbers[mid] == target
    return mid
  elsif numbers[mid] > target
    return binary_search(numbers[0..mid-1], target)
  else
    return binary_search(numbers[mid+1..-1], target)
  end
end